common/meparser.cpp

```diff
@@ -103,88 +103,135 @@ USTATUS MeParser::parseMeRegionBody(const UModelIndex & index)
-                            ptHeader->FlashLayout,
-                            ptHeader->FitcMajor, ptHeader->FitcMinor, ptHeader->FitcHotfix, ptHeader->FitcBuild,
-                            ptHeader->Checksum) + (ptHeader->Checksum == calculated ? UString("valid") : usprintf("invalid, should be %02Xh", calculated));
-    
+    UString info;
+
+    switch (pt1Header->HeaderVersion) {
+        case 0x10:
+            /* FALLTHROUGH */
+        case 0x20:
+            info = usprintf("Full size: %Xh (%u)\nHeader size: %Xh (%u)\nBody size: %Xh (%u)\nNumber of entries: %u\n"
+                            "Header version: %02Xh\nEntry version: %02Xh\nHeader length: %02Xh\n"
+                            "Ticks to add: %04Xh\nTokens to add: %04Xh\nUMA size: %Xh\nFlash layout: %Xh\n"
+                            "FITC version: %u.%u.%u.%u\nChecksum: %02Xh, ",
+                       ptSize, ptSize, header.size(), header.size(), ptBodySize, ptBodySize, pt1Header->NumEntries,
+                       pt1Header->HeaderVersion, pt1Header->EntryVersion, pt1Header->HeaderLength,
+                       pt1Header->TicksToAdd, pt1Header->TokensToAdd, pt1Header->UmaSize, pt1Header->FlashLayout,
+                       pt1Header->FitcMajor, pt1Header->FitcMinor, pt1Header->FitcHotfix, pt1Header->FitcBuild, pt1Header->Checksum);
+            info += (msgInvalidPtHeaderChecksum ? usprintf("invalid, should be %02Xh", chksumCalculated) : UString("valid"));
+            break;
+        case 0x21:
+            {
+                const FPT_HEADER21* pt2Header = (const FPT_HEADER21*)pt1Header;
+                info = usprintf("Full size: %Xh (%u)\nHeader size: %Xh (%u)\nBody size: %Xh (%u)\nNumber of entries: %u\n"
+                                "Header version: %02Xh\nEntry version: %02Xh\nHeader length: %02Xh\n"
+                                "Flags: %02Xh\nTicks to add: %04Xh\nTokens to add: %04Xh\nSPS Flags: %Xh\n"
+                                "FITC version: %u.%u.%u.%u\nChecksum: %08Xh, ",
+                        ptSize, ptSize, header.size(), header.size(), ptBodySize, ptBodySize, pt2Header->NumEntries,
+                        pt2Header->HeaderVersion, pt2Header->EntryVersion, pt2Header->HeaderLength,
+                        pt2Header->Flags, pt2Header->TicksToAdd, pt2Header->TokensToAdd, pt2Header->SPSFlags,
+                        pt2Header->FitcMajor, pt2Header->FitcMinor, pt2Header->FitcHotfix, pt2Header->FitcBuild, pt2Header->Checksum);
+                info += (msgInvalidPtHeaderChecksum ? usprintf("invalid, should be %08Xh", chksumCalculated) : UString("valid"));
+            }
+            break;
+        default:
+            info = usprintf("%s: XXX: unknown fpt header version - not yet!", __FUNCTION__);
+            break;
+    }
+
     // Add tree item
     index = model->addItem(0, Types::FptStore, 0, name, UString(), info, header, body, UByteArray(), Fixed, parent);
-    
+
     // Show messages
-    if (msgInvalidPtHeaderChecksum) {
```

Comment:
Please add this back.

Reply:
The message was moved up (lines around 170)